### PR TITLE
JENKINS-38096: add in authorization checks for multibranch workflow jobs

### PR DESCRIPTION
**Pull request by david-crowder**
This is my attempt at fulfilling:
https://issues.jenkins-ci.org/browse/JENKINS-38096

Tested against an enterprise Github instance.
Definitely open to suggestions on this. The jira indicated that this would be welcome functionality (and I wanted it) so I took a crack at it, despite not exactly being a java dev and by no means being a jenkins plugin developer. The changes to the inheritance structure in the workflow jobs made things much more confusing than I thought they would be. 

I appreciate any feedback.
Thanks!

### CODE REVIEW

**Comment by samrocketman** [comment]
Can you rebase off of the latest master?

**Comment by samrocketman** [comment]
@stephenc would you mind taking a look at this?

**Comment by samrocketman** [comment]
@jglick or @abayer since I'm unfamiliar with multibranch pipeline code do one of you mind reviewing this PR before I merge it?

I ask you because you're top contributors in addition to @stephenc  https://github.com/jenkinsci/workflow-multibranch-plugin

**Comment by stephenc** [comment]
It'll probably be Monday before I get a chance to take a look

File: src/main/java/org/jenkinsci/plugins/GithubAuthorizationStrategy.java [hunk 1]
```diff
@@ -84,11 +90,20 @@ public ACL getRootACL() {
     }
 
     @Nonnull
+    public ACL getACL(@Nonnull AbstractItem item) {
+        if(item instanceof WorkflowMultiBranchProject) {
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/GithubAuthorizationStrategy.java]: Should be Multibranch not workflow specific type

**Comment by samrocketman** [comment]
Thanks for taking the time.  I'll wait at least a week or so before merging this.

**Comment by david-crowder** [comment]
Sorry for the lack of updates, I am working on refactoring to be more generic as suggested. Will try and get something out this weekend.

File: src/main/java/org/jenkinsci/plugins/GithubRequireOrganizationMembershipACL.java [hunk 2]
```diff
@@ -270,18 +276,34 @@ public boolean hasRepositoryPermission(GithubAuthenticationToken authenticationT
...
     private String getRepositoryName() {
         String repositoryName = null;
-        SCM scm = this.project.getScm();
-        if (scm instanceof GitSCM) {
-            GitSCM git = (GitSCM)scm;
+        String repoUrl = null;
+        Describable scm = null;
+        if (this.item instanceof WorkflowJob) {
+            WorkflowMultiBranchProject project = (WorkflowMultiBranchProject) item.getParent();
+            scm = project.getSCMSources().get(0);
```
Comment (david-crowder) [inline src/main/java/org/jenkinsci/plugins/GithubRequireOrganizationMembershipACL.java]: Not sure what I was thinking here. I did a refactor to try and clean things up a bit, but I'm pretty sure this doesn't do what I was intending. Working to get that sorted now.

Reading over the dev guide (https://github.com/jenkinsci/pipeline-plugin/blob/master/DEVGUIDE.md#historical-background) it implies that you should rely on SCMTriggerItem to ascertain what sources a job pulls from, but that doesn't seem to be a holistic solution given that you can have jobs that don't have triggers defined. Unless I am missing something there.

**Comment by david-crowder** [comment]
I tried to find a way to do general checks on workflow jobs but I really have no idea the best way to do that, so I just locked the logic down to workflow jobs that are have a branch property (which was the original ask on the jira)... I'm not sure if this is better or worse. I also updated to use MultiBranchProject as suggested. At this point I really don't have any ideas on how to implement this in a more generic way. So any feedback on that would be greatly appreciated.

File: src/main/java/org/jenkinsci/plugins/GithubAuthorizationStrategy.java (hunk 1 above)
Comment (david-crowder) [inline src/main/java/org/jenkinsci/plugins/GithubAuthorizationStrategy.java]: Updated to use MultiBranch: https://github.com/jenkinsci/github-oauth-plugin/pull/78/commits/15233fb3db7ccaea0b845fadca192581e52f95a4

**Comment by david-crowder** [comment]
The build appears to have failed, but I am unable to reproduce that failure locally.

**Comment by david-crowder** [comment]
@stephenc, sorry to bug you, but did you have any further suggestions? It is still tightly tied to the multibranch project, but I really don't see a way around that :/

**Comment by david-crowder** [comment]
@samrocketman is there something I can do/change to help push this along?

**Comment by samrocketman** [comment]
Hey David, for this specific feature/fix it's outside of my area of expertise.  However, I'm willing to merge it after some compatibility testing of my own.  Others reviewed it and you appeared to address their feedback.

**Comment by samrocketman** [comment]
@stephenc one last double check and approval?

**Comment by samrocketman** [comment]
test this please

**Comment by samrocketman** [comment]
Close/reopen to get jenkins to revalidate.

**Comment by david-crowder** [comment]
Thanks @samrocketman!

File: pom.xml [hunk 3]
```diff
@@ -112,6 +112,16 @@
             <version>1.7</version>
         </dependency>
 
+        <dependency>
+            <groupId>org.jenkins-ci.plugins.workflow</groupId>
+            <artifactId>workflow-multibranch</artifactId>
```
Comment (jglick) [inline pom.xml]: This dependency should not be here I think. Maybe you were looking for `SCMSourceOwner`?

File: pom.xml [hunk 4]
```diff
@@ -112,6 +112,16 @@
...
+            <version>2.9</version>
+        </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>github-branch-source</artifactId>
```
Comment (jglick) [inline pom.xml]: And this should be an `optional` dependency.

**Pull request merged**